pkg/parse/opts.go

```diff
@@ -52,56 +52,40 @@ type Options struct {
 	// SyncName is the name of the RootSync or RepoSync object.
 	SyncName string
 
-	// StatusUpdatePeriod is how long the Parser waits between updates of the
-	// sync status, to account for management conflict errors from the Remediator.
-	StatusUpdatePeriod time.Duration
+	// Scope defines the scope of the reconciler, either root or namespaced.
+	Scope declared.Scope
 
-	// DiscoveryInterface is how the Parser learns what types are currently
+	// DiscoveryClient is how the Parser learns what types are currently
 	// available on the cluster.
-	DiscoveryInterface discovery.ServerResourcer
+	DiscoveryClient discovery.ServerResourcer
 
 	// Converter uses the DiscoveryInterface to encode the declared fields of
 	// objects in Git.
 	Converter *declared.ValueConverter
 
-	// mux prevents status update conflicts.
-	mux sync.Mutex
-
-	// RenderingEnabled indicates whether the hydration-controller is currently
-	// running for this reconciler.
-	RenderingEnabled bool
-
 	// WebhookEnabled indicates whether the Webhook is currently enabled
 	WebhookEnabled bool
 
-	// Files lists Files in the source of truth.
-	Files
-	// Updater mutates the most-recently-seen versions of objects stored in memory.
-	Updater
+	// DeclaredResources is the set of valid source objects, managed by the
+	// Updater and shared with the Parser & Remediator.
+	// This is used by the Parser to validate that CRDs can only be removed from
+	// the source when all of its CRs are removed as well.
+	DeclaredResources *declared.Resources
 }
 
-// Parser represents a parser that can be pointed at and continuously parse a source.
-type Parser interface {
-	parseSource(ctx context.Context, state *sourceState) ([]ast.FileObject, status.MultiError)
-	ReconcilerStatusFromCluster(ctx context.Context) (*ReconcilerStatus, error)
-	setSourceStatus(ctx context.Context, newStatus *SourceStatus) error
-	setRenderingStatus(ctx context.Context, oldStatus, newStatus *RenderingStatus) error
-	SetSyncStatus(ctx context.Context, newStatus *SyncStatus) error
-	options() *Options
-	// SyncErrors returns all the sync errors, including remediator errors,
-	// validation errors, applier errors, and watch update errors.
-	SyncErrors() status.MultiError
-	// K8sClient returns the Kubernetes client that talks to the API server.
-	K8sClient() client.Client
-	// setRequiresRendering sets the requires-rendering annotation on the RSync
-	setRequiresRendering(ctx context.Context, renderingRequired bool) error
-	setSourceAnnotations(ctx context.Context, commit string) error
-}
+// ReconcilerOptions holds configuration for the reconciler.
+type ReconcilerOptions struct {
```

Comment:
I understand it's a little confusing. That's because these "options" are being used to simplify construction but not being used as part of a builder pattern. They're being directly composed into the object they're supposed to be configuring. This is sort of an anti-pattern, but I chose to keep it for this pass to reduce the number of changes a little. 

So the answer is, if the parser impls use it, I left it in the parse.Options. Everything else I moved into the ReconcilerOptions. And then I made the ReconcilerOptions compose the Options, simply to avoid needing to copy values from one to the other and keep them in sync. 

There's definitely still some room for improvement in future passes. Like maybe just removing the options structs and building the objects directly. Or with a builder, for the parsers, which have multiple impls.

Reply:
This is largely just a dependency injection problem. It wasn't the primary focus of this refactor to clean that up. But I could spend a little more time on it if you want it in the same PR.